Handle conditional dependency resolution in Gradle

glefloch opened the pull request:
This branch adds support for conditional dependencies in gradle. This is still a work in progress. 
What it does: 

- looks for extension 
- looks for conditional dependencies in extension
- resolve conditional dependencies and look for dependency condition 
- validate condition 
- add dependency to compileClasspath if condition is met

What needs to be done: 

- Handle recursive resolution (conditional dependency in conditional dependency)
- Handle explicit dependency exclusion

I haven't used feature variant as it was not needed but it would have worked. 

@aloubyansky WDYT?

glefloch committed:
Handle conditional dependency resolution in Gradle

Signed-off-by: Guillaume Le Floch <glfloch@gmail.com>

glefloch commented:
@aloubyansky I will check failed tests but you already can have a look at it.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/dependency/ConditionalDependenciesEnabler.java:
```diff
@@ -0,0 +1,193 @@
+        extension.importConditionalDependency(project, dependency);
+    }
+
+    private ExtensionDependency getExtensionInfoOrNull(ResolvedArtifact artifact, File artifactFile,
```
It looks like the `artifactFile` is always `artifact.getFile()`, so the only method argument needed here is `artifact`.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/dependency/ConditionalDependenciesEnabler.java:
```diff
@@ -0,0 +1,193 @@
+        Dependency constraint = null;
+        if (extensionProperties.containsKey(BootstrapConstants.DEPENDENCY_CONDITION)) {
+            constraint = DependencyUtils.create(dependencies,
+                    extensionProperties.getProperty(BootstrapConstants.DEPENDENCY_CONDITION));
```
The condition may include multiple artifacts (artifact keys).

glefloch replied:
oh ok, I will create a list here. And what happen when no condition is declared? right now, the conditional dependency won't be importer, should we?

aloubyansky replied:
In the Maven impl, no condition is currently treated as being "satisfied".

glefloch replied:
ok, I will update that then.

glefloch replied:
just to make sure, if there is multiple conditions, all condition should be met, or if one is met it's ok?

aloubyansky replied:
To be more precise, a single condition may include one or more artifact keys. All the artifact keys must be found on the classpath for the condition to be satisfied.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/dependency/ConditionalDependenciesEnabler.java:
```diff
@@ -0,0 +1,193 @@
+        }
+    }
+
+    private Configuration createConditionalDependenciesConfiguration(Project project, Configuration existingDeps,
```
The `project` is an instance variable, right? I.e. it doesn't have to be passed in as an arg?

glefloch replied:
right!

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/dependency/ConditionalDependenciesEnabler.java:
```diff
@@ -0,0 +1,193 @@
+        return possibleVariant;
+    }
+
+    private void resolveConditionalDependencies(DependencyHandler dependencies,
```
`dependencies` seems to be always `project.getDependencies()`? Doesn't have to be passed as a method argument?

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/dependency/ExtensionDependency.java:
```diff
@@ -0,0 +1,101 @@
+            dep.exclude(DependencyUtils.exclusionRuleToMap(excludeRule));
+        }
+
+        project.getDependencies().add(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME, dep);
```
Conditional dependencies shouldn't be added as direct project dependencies, they should be added as capability dependencies to the extensions that declare conditional dependencies instead. Do you think you can refactor it that way?

glefloch replied:
I don't think we can, each dependency can require one capability (from one I understood), by default, this requires a capability with the same name as the dependency. To require a new capability, we have to declare a new dependency. In the test example, we will end up with something similar too : 

    implementation("org.acme:quarkus-ext-a:1.0-SNAPSHOT")
    implementation("org.acme:quarkus-ext-a:1.0-SNAPSHOT") {
         capabilities {
            requireCapability("org.acme:quarkus-ext-b-feature")
        }
    }

aloubyansky replied:
I thought we could modify metadata of an existing dependency, couldn't we?

glefloch replied:
we can, that what we do when adding a variant. I will try to require that capability there too.

glefloch replied:
You were right, actually, this can be simplified to updating existing variant and adding a dependency
 to it.

glefloch commented:
CI is finally happy :) I addressed your comments @aloubyansky if you want to have an other look at it.

aloubyansky commented:
Could you please incorporate a test I committed here? https://github.com/aloubyansky/quarkus/commit/31a867b259e471897c9de0de0dcedc4a3a60abb3 It will most probably fail, it was taking too long to run so stopped it.

glefloch commented:
yes sure, I will add it.

aloubyansky commented:
Just in case, i added this scenario to a Maven based test here https://github.com/aloubyansky/quarkus/commit/5ef786ea6bebfa02c9a795e3d37907d6fce02496 to make sure, it is supported on the Maven side too.

glefloch commented:
@aloubyansky I pushed the scenario tow test. It's ok locally. I updated the expected result to match the maven test.

aloubyansky commented:
Could you also add this adjustment to the test? https://github.com/aloubyansky/quarkus/commit/69a2651002540d1b63a1705e3b2df81da50baf2e This one actually fails.

aloubyansky commented:
Hold on, sorry, I messed it up.

aloubyansky commented:
@glefloch could you check this adjustment? https://github.com/aloubyansky/quarkus/commit/e5f8a37d021034f9968a678131d56c1634b234bc

glefloch commented:
Yes sure

glefloch commented:
@aloubyansky I updated the test. It showed an error, conditional-dependency were resolved based on available dependencies. I wasn't retrying to resolve previous conditional-dependency after adding some extension. I fixed it and tests are ok.

aloubyansky commented:
@glefloch here is the next test adjustment https://github.com/aloubyansky/quarkus/commit/5f7cc296252433cf70e592698ce7b9443826a9f8
If we could generate those extension projects using some builder api, it would be less error-prone.

glefloch commented:
yes I agree this is not really easy to update those projects

glefloch commented:
hmmm I lost `o` and `p` I will add some logs to understand why they are gone

glefloch commented:
Ok, there was an issue in the code I added yesterday. I fixed it and the test is ok :)

aloubyansky commented:
Now let's make sure the deployment classpath is complete https://github.com/aloubyansky/quarkus/commit/153a8b28721e2f9e01e6d9c0ea331460b471dc5d

aloubyansky commented:
Basically, the configuration with the conditional deps activated should be used in that method building the model instead of the original configuration.

glefloch commented:
@aloubyansky I created a `quarkusExtensionConfig` that will contains all extensions. This is then used to build the model. Once this branch is ok, I think I will add a rule to replace extensions by its deployment artifact. (can by assume that all deployment artifact name is the name of the artifact suffixed by `-deployment` ? )

aloubyansky commented:
While the -deployment suffix is pretty much a rule, it's still officially only the default convention that may be re-configured. The other issue is that the -deployment suffix is not exclusive to the extension deployment artifacts. Other non-extension artifacts may happen to have this suffix as well.

glefloch commented:
@aloubyansky deployment classpath is now ok :)

aloubyansky commented:
Next one https://github.com/quarkusio/quarkus/commit/d8a4f280811acbf0c8e931fb2167608487306781, actually if you change the dependency condition of `N` to `K` it should already break.

glefloch commented:
it should be good now :)

glefloch commented:
@aloubyansky tests failures are not related to this development. Some deps are wrongly resolved in tests, I pushed an other branch to fix those tests.

aloubyansky commented on integration-tests/gradle/src/test/resources/conditional-dependencies/gradle.properties:
```diff
@@ -0,0 +1,4 @@
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPlatformGroupId=io.quarkus
+quarkusPlatformVersion=999-SNAPSHOT
```
999-SNAPSHOT won't be available during the release process. The version should be added in the test initialization like in other gradle tests (or by resources filtering, if that's an option).

glefloch replied:
Oh right, this is the case, I think it was for local tests

aloubyansky commented on integration-tests/gradle/update-dependencies.sh:
```diff
@@ -37,8 +37,8 @@ echo ''
-              grep -Eo 'quarkus-[a-z0-9-]+' | \
-              sed -e '/-deployment/d' -e '/quarkus-bom/d' | sort | uniq)
+              grep -Eo ':quarkus-[a-z0-9-]+' | \
+              sed -e '/-deployment/d' -e '/quarkus-bom/d' -e 's/://' | sort | uniq)
```
@famod could you please check this specific change?

glefloch replied:
@famod, we reference dependencies that are build by an other project so they should be tracked by this script.

famod replied:
Ok, I saw those `ext-quarkus-a` (b, c, etc.) test dependencies. I have not better suggestion given the required BSD compatibility.
Or maybe I have: _Not_ using `quarkus` in those deps is not an option?

glefloch replied:
Yes we could use something Elsa that quarkus in the name. How this is solved in the maven plugin module?

aloubyansky replied:
We could simply remove `quarkus` from the artifactIds, i.e. `ext-a`, etc.

glefloch replied:
Yes I will revert this change and update project names

aloubyansky commented on integration-tests/gradle/src/test/resources/conditional-dependencies/ext-quarkus-b/deployment/src/main/java/org.acme.quarkus.ext.a.deployment/AcmeQuarkusExtProcessor.java:
```diff
@@ -0,0 +1,16 @@
+
```
I've noticed the packages aren't laid out as directories on the filesystem, i.e. `src/main/java/org.acme.quarkus.ext.a.deployment/AcmeQuarkusExtProcessor.java ` was that on purpose?

glefloch replied:
Oh no, it's not, I will update that too.

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java:
```diff
@@ -80,6 +82,17 @@ private static Configuration classpathConfig(Project project, LaunchMode mode) {
+        return project.getConfigurations().create(QUARKUS_MODEL_CONFIGURATION)
+                .extendsFrom(project.getConfigurations()
+                        .getByName(ConditionalDependenciesEnabler.QUARKUS_EXTENSION_CONFIGURATION_NAME))
+                .extendsFrom(classpathConfiguration);
```
Is this making all the extensions (including the transitive ones) direct (top level) dependencies in this configuration?

glefloch replied:
Yes it is

aloubyansky replied:
This looks suspicious to me. Direct dependencies and their order is significant for version convergence. What was the reason to do it this way?

glefloch replied:
I did this because when we collect deployment dependencies, we only look for first level dependencies. Which does not include conditional dependencies.

aloubyansky replied:
Ah, I see. To not miss the deployment deps of the conditional deps. Ok, I need to reflect on that.

glefloch replied:
Exactly

glefloch replied:
I don't how we could preserve the resolution order

famod commented on integration-tests/gradle/src/test/resources/conditional-dependencies/ext-quarkus-a/deployment/build.gradle:
```diff
@@ -0,0 +1,22 @@
+
+publishing {
+    publications {
+        mavenJave(MavenPublication) {
```
Just saw this by chance:
```suggestion
        mavenJava(MavenPublication) {
```
But then again I have no clue what I'm talking about. 😄

glefloch replied:
nice catch!

aloubyansky commented on devtools/gradle/src/main/java/io/quarkus/gradle/dependency/ConditionalDependenciesEnabler.java:
```diff
@@ -0,0 +1,207 @@
+
+        Configuration resolvedExtensions = DependencyUtils.duplicateConfiguration(project,
+                project.getConfigurations().getByName(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME),
+                project.getConfigurations().getByName(QuarkusPlugin.DEV_MODE_CONFIGURATION_NAME));
```
Do we always want the dev mode only dependencies in this configuration? Won't they leak into the application's lib dir?

aloubyansky replied:
Could you clarify why this is needed including also the `exposeExtensionConfiguration`? I am probably missing something but it looks like the `configurationCopy` at this point is the proper runtime config. Is it not?

aloubyansky replied:
I mean the `configurationCopy` before this line, right after `resolveConditionalDependencies`.

glefloch replied:
This is needed by the model builder. This make sure all extensions are declared on a first level.

glefloch replied:
The `configurationCopy` is correct but extensions are not first level module. We could keep it this way (without exposing the extension configuration) but we would need to update the builder to look for extensions in all the entire runtime classpath

aloubyansky replied:
AFAIU, it's about collecting the conditional deployment deps. Otherwise, with the `configurationCopy` (which should probably be renamed to something more appropriate, btw) it would work like before.
Would you mind pushing it a little further? I was never happy with the way we resolve the deployment classpath and now with the conditional deps it looks like we have no choice but do it right.
Having identified all the runtime extension artifacts on the classpath (including the conditional ones), you should also know all their corresponding `deployment-artifact`s (they are in the same `extension-descriptor.properties` file).
Now, if you apply dependency substitution to the `configurationCopy` replacing all the runtime extension artifacts with the deployment ones, you'll get the proper application build classpath. But keep the `configurationCopy` as the source for application runtime `lib`. Does it make sense? Are you not exhausted yet? :)

glefloch replied:
yes it totally make sense. I will do that. 

not exhausted yet :)

glefloch commented:
@aloubyansky I pushed all updates, there are some tests failures on ci but locally it looks ok. I will investigate.
